File: Phase2.py
def print_representations(number):
    # Check for negative number for unsigned format
    if number < 0:
        print("Error: Negative numbers cannot be represented in the unsigned format.")
    else:
        # Unsigned binary
        unsigned_binary = f'{number & 0xFFFFFFFF:032b}'
        unsigned_hex = f'{number & 0xFFFFFFFF:08X}'
        unsigned_octal = f'{number & 0xFFFFFFFF:011o}'
        print(f'Unsigned binary:   {unsigned_binary}')1
        print(f'Unsigned hex:      {unsigned_hex}')
        print(f'Unsigned octal:    {unsigned_octal}')

    # Signed binary
    signed_binary = f'{number if number >= 0 else (number + (1 << 32)):032b}'
    signed_hex = f'{number if number >= 0 else (number + (1 << 32)):08X}'
    signed_octal = f'{number if number >= 0 else (number + (1 << 32)):011o}'
    print(f'Signed binary:     {signed_binary}')
    print(f'Signed hex:        {signed_hex}')
    print(f'Signed octal:      {signed_octal}')

    # One's complement
    ones_complement = ''.join('1' if bit == '0' else '0' for bit in signed_binary)
    ones_complement_hex = f'{int(ones_complement, 2):08X}'
    ones_complement_octal = f'{int(ones_complement, 2):011o}'
    print(f"One's complement:  {ones_complement}")
    print(f"One's comp. hex:   {ones_complement_hex}")
    print(f"One's comp. octal: {ones_complement_octal}")

    # Two's complement
    twos_complement = f'{(number if number >= 0 else ((number + (1 << 32)) + 1) & 0xFFFFFFFF):032b}'
    twos_complement_hex = f'{int(twos_complement, 2):08X}'
    twos_complement_octal = f'{int(twos_complement, 2):011o}'
    print(f"Two's complement:  {twos_complement}")
    print(f"Two's comp. hex:   {twos_complement_hex}")
    print(f"Two's comp. octal: {twos_complement_octal}")

# Example usage
num = int(input("Enter a number: "))
print_representations(num)

File: Phase2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void int_to_bin(int num, char* output, int output_size) {
    output[output_size - 1] = '\0';
    for (int i = output_size - 2; i >= 0; i--) {
        output[i] = (num & 1) ? '1' : '0';
        num >>= 1;
    }
}

void print_representations(int number) {
    char buffer[33]; // Buffer to hold binary representations

    // Check for negative number for unsigned format
    if (number < 0) {
        printf("Error: Negative numbers cannot be represented in the unsigned format.\n");
    } else {
        // Unsigned binary
        unsigned int unsigned_number = (unsigned int) number;
        int_to_bin(unsigned_number, buffer, 33);
        printf("Unsigned binary:   %s\n", buffer);
        printf("Unsigned hex:      %08X\n", unsigned_number);
        printf("Unsigned octal:    %011o\n", unsigned_number);
    }

    // Signed binary
    int_to_bin(number, buffer, 33);
    printf("Signed binary:     %s\n", buffer);
    printf("Signed hex:        %08X\n", number);
    printf("Signed octal:      %011o\n", number);

    // One's complement
    for (int i = 0; i < 32; i++) {
        buffer[i] = (buffer[i] == '0') ? '1' : '0';
    }
    unsigned int ones_complement_value = strtoul(buffer, NULL, 2);
    printf("One's complement:  %s\n", buffer);
    printf("One's comp. hex:   %08X\n", ones_complement_value);
    printf("One's comp. octal: %011o\n", ones_complement_value);

    // Two's complement
    unsigned int twos_complement_value;
    if (number < 0) {
        twos_complement_value = (unsigned int)(number + (1 << 31)) + 1; // Adjust to correct left shift
        int_to_bin(twos_complement_value, buffer, 33);
        printf("Two's complement:  %s\n", buffer);
        printf("Two's comp. hex:   %08X\n", twos_complement_value);
        printf("Two's comp. octal: %011o\n", twos_complement_value);
    } else {
        int_to_bin(number, buffer, 33);
        printf("Two's complement:  %s\n", buffer);
        printf("Two's comp. hex:   %08X\n", number);
        printf("Two's comp. octal: %011o\n", number);
    }
}

int main() {
    int num;
    printf("Enter a number: ");
    scanf("%d", &num);
    print_representations(num);
    return 0;
}
